apps/extension/src/content-scripts/injected-penumbra-global.ts

```diff
@@ -5,54 +5,101 @@
- * the page. Imports, includes, and packages your webapp depends on could all
- * mutate or preempt the global. User-agent injections like userscripts or other
- * content scripts could interfere or intercept connections.
+ * the page. Imports, requires, includes, script tags, packages your webapp
+ * depends on, userscripts, or other extensions' content scripts could all
+ * mutate or preempt this, and all have the power to interfere or intercept
+ * connections.
  */
 
 import { PenumbraInjection, PenumbraRequestFailure, PenumbraSymbol } from '@penumbra-zone/client';
-import { isPraxFailureMessageEvent, isPraxPortMessageEvent, PraxMessage } from './message-event';
+
+import {
+  isPraxFailureMessageEvent,
+  isPraxPortMessageEvent,
+  PraxMessage,
+  unwrapPraxMessage,
+} from './message-event';
 
 import { PraxConnection } from '../message/prax';
 
-const request = Promise.withResolvers<void>();
-
-// this is just withResolvers, plus a sync-queryable state attribute
-const connection = Object.assign(Promise.withResolvers<MessagePort>(), { state: false });
-connection.promise.then(
-  () => {
-    connection.state = true;
-    request.resolve();
-  },
-  () => {
-    connection.state = false;
-    request.reject();
-  },
-);
-
-// this resolves the connection promise when the isolated port script indicates
-const connectionListener = (msg: MessageEvent<unknown>) => {
-  if (isPraxPortMessageEvent(msg) && msg.origin === window.origin) {
-    // @ts-expect-error - ts can't understand the injected string
-    const praxPort: unknown = msg.data[PRAX];
-    if (praxPort instanceof MessagePort) connection.resolve(praxPort);
+type PromiseSettledResultStatus = PromiseSettledResult<unknown>['status'];
+
+class PraxInjection {
+  private static singleton?: PraxInjection = new PraxInjection();
+
+  public static get penumbra() {
+    return PraxInjection.singleton!.injection;
+  }
+
+  private manifestUrl = `${PRAX_ORIGIN}/manifest.json`;
+  private _request = Promise.withResolvers<void>();
+  private _connect = Promise.withResolvers<MessagePort>();
+  private _disconnect = Promise.withResolvers<void>();
+
+  private connectState?: PromiseSettledResultStatus;
+  private requestState?: PromiseSettledResultStatus;
+  private disconnectState?: PromiseSettledResultStatus;
+
+  private injection: Readonly<PenumbraInjection> = Object.freeze({
+    disconnect: () => this.endConnection(),
+    connect: () => (this.state() !== false ? this._connect.promise : this.connectionFailure),
+    isConnected: () => this.state(),
+    request: () => this.postRequest(),
+    manifest: String(this.manifestUrl),
+  });
+
+  private constructor() {
+    if (PraxInjection.singleton) return PraxInjection.singleton;
+
+    window.addEventListener('message', this.connectionListener);
+
+    void this._connect.promise
+      .then(
+        () => (this.connectState ??= 'fulfilled'),
+        () => (this.connectState ??= 'rejected'),
```

Comment:
Why do we use `??=` here? Shouldn't we respect the latest value?

Reply:
these state variables are for tracking promise state synchronously. a promise can only fulfill or reject once, so the variables may only be set once.